File: server/src/main/com/broll/gainea/server/core/bot/impl/BotCard.java
package com.broll.gainea.server.core.bot.impl;

import com.broll.gainea.net.NT_Action;
import com.broll.gainea.net.NT_Action_Card;
import com.broll.gainea.net.NT_Reaction;
import com.broll.gainea.server.core.bot.BotOptionalAction;
import com.broll.gainea.server.core.bot.CardOption;
import com.broll.gainea.server.core.bot.strategy.ICardStrategy;
import com.broll.gainea.server.core.cards.Card;

public class BotCard extends BotOptionalAction<NT_Action_Card, CardOption> {

    @Override
    protected void react(NT_Action_Card action, NT_Reaction reaction) {
        BotSelect select = (BotSelect) handler.getActionHandler(BotSelect.class);
        select.clearSelections();
        getSelectedOption().getSelectOptions().forEach(select::nextChooseOption);
    }

    @Override
    public Class<? extends NT_Action> getActionClass() {
        return NT_Action_Card.class;
    }

    @Override
    public CardOption score(NT_Action_Card action) {
        BotSelect select = (BotSelect) handler.getActionHandler(BotSelect.class);
        Card card = bot.getCardHandler().getCards().stream().filter(it -> it.getId() == action.cardId).findFirst().orElse(null);
        if (card instanceof ICardStrategy) {
            return ((ICardStrategy) card).strategy(strategy, select);
        }
        return null;
    }

}

File: server/src/main/com/broll/gainea/server/core/bot/impl/BotCard.kt
package com.broll.gainea.server.core.bot.impl

import com.broll.gainea.net.NT_Action
import com.broll.gainea.net.NT_Action_Card
import com.broll.gainea.net.NT_Reaction
import com.broll.gainea.server.core.bot.BotOptionalAction
import com.broll.gainea.server.core.bot.CardOption
import com.broll.gainea.server.core.bot.strategy.ICardStrategy

class BotCard : BotOptionalAction<NT_Action_Card, CardOption>() {
    override fun react(action: NT_Action_Card, reaction: NT_Reaction) {
        val select = handler.getActionHandler(BotSelect::class.java)
        select.clearSelections()
        selectedOption!!.getSelectOptions().forEach { select.nextChooseOption(it) }
    }

    override val actionClass: Class<out NT_Action>
        get() = NT_Action_Card::class.java

    override fun score(action: NT_Action_Card): CardOption? {
        val select = handler.getActionHandler(BotSelect::class.java)
        val card = bot.cardHandler.cards.find { it.id == action.cardId }
        return if (card is ICardStrategy) {
            (card as ICardStrategy).strategy(strategy, select)
        } else null
    }
}
